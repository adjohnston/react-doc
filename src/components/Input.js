import React from 'react'
import PropTypes from 'prop-types'
import classString from 'src/helpers/class-string'
import transformValue from 'src/helpers/transform-value'

const handleInputOnChange = (onChange, name, type) => ({ target }) => {
  return onChange(name, transformValue(
    (type === 'bool')
      ? target.checked
      : target.value
  ), type)
}

const Input = ({name, type, onChange, value}) => {
  const handleOnChange = handleInputOnChange(onChange, name, type)

  let inputElement
  if (type === 'string') {
    inputElement = (
      <input
        type='text'
        value={value}
        className={classString('__input')}
        onChange={handleOnChange} />
    )
  }

  if (type === 'number') {
    inputElement = (
      <input
        type='number'
        value={value}
        className={classString('__input')}
        onChange={handleOnChange} />
    )
  }

  if (type === 'bool') {
    inputElement = (
      <input
        type='checkbox'
        checked={value}
        className={classString('__checkbox')}
        onChange={handleOnChange} />
    )
  }

  return inputElement || null
}

Input.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool
  ])
}

export { handleInputOnChange }
export default Input
