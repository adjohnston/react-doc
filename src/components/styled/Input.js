import styled from 'styled-components'

export const InputText = styled.input`
  width: 100%;
  height: 2rem;
  padding: 0 .5rem;
  font: normal 1rem/2 arial;
`
