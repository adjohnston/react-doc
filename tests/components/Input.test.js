import React from 'react'
import renderer from 'react-test-renderer'
import Input from 'src/components/Input'

const mockInputStringProps = {
  name: 'className',
  type: 'string',
  value: 'button',
  onChange: () => {}
}

const mockCheckboxBoolProps = {
  name: 'isActive',
  type: 'bool',
  value: false,
  onChange: () => {}
}

test('renders correctly', () => {
  const tree = renderer
    .create(<Input {...mockInputStringProps} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

test('renders input type of checkbox given a type of bool', () => {
  const tree = renderer
    .create(<Input {...mockCheckboxBoolProps} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

test('renders input type of checkbox given a type of bool and has attribute checked', () => {
  const {
    value,
    ...restProps
  } = mockCheckboxBoolProps

  const tree = renderer
    .create(<Input {...restProps} value={true} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})
