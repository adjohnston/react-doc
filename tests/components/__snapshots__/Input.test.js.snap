// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<input
  className="stncl__input"
  onChange={[Function]}
  type="text"
  value="button"
/>
`;

exports[`renders input type of checkbox given a type of bool 1`] = `
<input
  checked={false}
  className="stncl__checkbox"
  onChange={[Function]}
  type="checkbox"
/>
`;

exports[`renders input type of checkbox given a type of bool and has attribute checked 1`] = `
<input
  checked={true}
  className="stncl__checkbox"
  onChange={[Function]}
  type="checkbox"
/>
`;
